sqlglot/optimizer/simplify.py

```diff
@@ -407,28 +409,49 @@ def absorb_and_eliminate(expression, root=True):
     if isinstance(expression, exp.Connector) and (root or not expression.same_parent):
         kind = exp.Or if isinstance(expression, exp.And) else exp.And
 
-        for a, b in itertools.permutations(expression.flatten(), 2):
-            if isinstance(a, kind):
-                aa, ab = a.unnest_operands()
-
-                # absorb
-                if is_complement(b, aa):
-                    aa.replace(exp.true() if kind == exp.And else exp.false())
-                elif is_complement(b, ab):
-                    ab.replace(exp.true() if kind == exp.And else exp.false())
-                elif (set(b.flatten()) if isinstance(b, kind) else {b}) < set(a.flatten()):
-                    a.replace(exp.false() if kind == exp.And else exp.true())
-                elif isinstance(b, kind):
-                    # eliminate
-                    rhs = b.unnest_operands()
-                    ba, bb = rhs
-
-                    if aa in rhs and (is_complement(ab, ba) or is_complement(ab, bb)):
-                        a.replace(aa)
-                        b.replace(aa)
-                    elif ab in rhs and (is_complement(aa, ba) or is_complement(aa, bb)):
-                        a.replace(ab)
-                        b.replace(ab)
+        ops = tuple(expression.flatten())
+
+        complements = set()
+        subops = defaultdict(list)
+        pairs = defaultdict(list)
```

Comment:
The previous algorithm was a bit easier to follow, is it possible to leave a comment explaining the usecase of these collections?

Reply:
Added some comments